File: packages/luciq_flutter/android/src/main/java/ai/luciq/flutter/util/RunCatching.java
package ai.luciq.flutter.util;

import android.util.Log;

/**
 * Defensive try/catch wrapper for native Pigeon API entry points so the SDK
 * never crashes the host app. Mirrors the Dart-side {@code runCatching}
 * helper at {@code lib/src/utils/run_catching.dart}.
 *
 * <p>Catches {@link Throwable} (including {@code Error} subclasses) — the
 * mandate is that an observability SDK must never be the cause of a crash,
 * which the Dart layer also enforces. Logs the failure to Logcat.
 */
public final class RunCatching {
    private static final String TAG = "Luciq";

    private RunCatching() {}

    @FunctionalInterface
    public interface ThrowingRunnable {
        void run() throws Throwable;
    }

    @FunctionalInterface
    public interface ThrowingSupplier<T> {
        T get() throws Throwable;
    }

    /** Runs {@code action}; logs and swallows anything thrown. */
    public static void runCatching(String method, ThrowingRunnable action) {
        try {
            action.run();
        } catch (Throwable t) {
            Log.e(TAG, method + " failed", t);
        }
    }

    /** Runs {@code action}; logs and returns {@code fallback} on throw. */
    public static <T> T runCatchingReturn(String method, T fallback, ThrowingSupplier<T> action) {
        try {
            return action.get();
        } catch (Throwable t) {
            Log.e(TAG, method + " failed", t);
            return fallback;
        }
    }
}

File: packages/luciq_flutter/lib/src/utils/run_catching.dart
import 'dart:async';

import 'package:luciq_flutter/src/utils/luciq_logger.dart';

const _tag = 'Luciq';

/// Runs a synchronous void [action], logging any thrown error to
/// [LuciqLogger] under [method] and swallowing it. Used so the SDK never
/// crashes the host app.
void runCatching(String method, void Function() action) {
  try {
    action();
  } catch (e, st) {
    LuciqLogger.I.e('$method failed: $e\n$st', tag: _tag);
  }
}

/// Async variant of [runCatching] for `Future<void>` methods.
Future<void> runCatchingAsync(
  String method,
  FutureOr<void> Function() action,
) async {
  try {
    await action();
  } catch (e, st) {
    LuciqLogger.I.e('$method failed: $e\n$st', tag: _tag);
  }
}

/// Async variant that returns a value, falling back to [fallback] on error.
Future<T> runCatchingReturn<T>(
  String method,
  FutureOr<T> Function() action, {
  required T fallback,
}) async {
  try {
    return await action();
  } catch (e, st) {
    LuciqLogger.I.e('$method failed: $e\n$st', tag: _tag);
    return fallback;
  }
}
